atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java

```diff
@@ -15,94 +15,95 @@
+import com.palantir.atlasdb.pue.SimpleCommitTimestampPutUnlessExistsTable;
+import com.palantir.atlasdb.transaction.encoding.CellEncodingStrategy;
 import com.palantir.atlasdb.transaction.encoding.TicketsEncodingStrategy;
 import com.palantir.atlasdb.transaction.encoding.TimestampEncodingStrategy;
+import com.palantir.atlasdb.transaction.encoding.TwoPhaseEncodingStrategy;
 import com.palantir.atlasdb.transaction.encoding.V1EncodingStrategy;
 import com.palantir.atlasdb.transaction.impl.TransactionConstants;
-import java.util.HashMap;
 import java.util.Map;
 
 public final class SimpleTransactionService implements EncodingTransactionService {
-    private final KeyValueService kvs;
-    private final TimestampEncodingStrategy encodingStrategy;
-    private final TableReference transactionsTable;
-
-    // The maximum key-value store timestamp (exclusive) at which data is stored
-    // in transaction table.
-    // All entries in transaction table are stored with timestamp 0
-    private static final long MAX_TIMESTAMP = 1L;
-    private final AsyncCellGetter immediateAsyncCellGetter;
-    private final AsyncCellGetter asyncCellGetter;
+    private final PutUnlessExistsTable<Long, Long> txnTable;
+    private final TimestampEncodingStrategy<?> encodingStrategy;
 
     private SimpleTransactionService(
-            KeyValueService kvs, TimestampEncodingStrategy encodingStrategy, TableReference transactionsTable) {
-        this.kvs = kvs;
+            PutUnlessExistsTable<Long, Long> txnTable, TimestampEncodingStrategy<?> encodingStrategy) {
         this.encodingStrategy = encodingStrategy;
-        this.transactionsTable = transactionsTable;
-        this.immediateAsyncCellGetter = startTsMap -> Futures.immediateFuture(kvs.get(transactionsTable, startTsMap));
-        this.asyncCellGetter = startTsMap -> kvs.getAsync(transactionsTable, startTsMap);
+        this.txnTable = txnTable;
     }
 
     public static SimpleTransactionService createV1(KeyValueService kvs) {
-        return new SimpleTransactionService(kvs, V1EncodingStrategy.INSTANCE, TransactionConstants.TRANSACTION_TABLE);
+        return createSimple(kvs, TransactionConstants.TRANSACTION_TABLE, V1EncodingStrategy.INSTANCE);
     }
 
     public static SimpleTransactionService createV2(KeyValueService kvs) {
-        return new SimpleTransactionService(
-                kvs, TicketsEncodingStrategy.INSTANCE, TransactionConstants.TRANSACTIONS2_TABLE);
+        return createSimple(kvs, TransactionConstants.TRANSACTIONS2_TABLE, TicketsEncodingStrategy.INSTANCE);
+    }
+
+    public static SimpleTransactionService createV3(KeyValueService kvs) {
+        if (kvs.getCheckAndSetCompatibility().consistentOnFailure()) {
+            return createSimple(kvs, TransactionConstants.TRANSACTIONS3_TABLE, TicketsEncodingStrategy.INSTANCE);
+        }
+        return createResilient(kvs, TransactionConstants.TRANSACTIONS3_TABLE, TwoPhaseEncodingStrategy.INSTANCE);
```

Comment:
actually: do we really need a transactions3 table? It is probably more efficient to just continue to use _transactions2, but access it differently (the timestamp partitioning map will ensure we read from it properly, and this saves us a lot of cassandra tables).